tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/CloneRepos.cs

```diff
@@ -0,0 +1,155 @@
+﻿using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+
+namespace Microsoft.DotNet.SourceBuild.Tasks
+{
+    public class CloneRepos : Microsoft.Build.Utilities.Task
+    {
+        [Required]
+        public string SourceRoot { get; set; }
+
+        [Required]
+        public string SpecFile { get; set; }
+
+        public override bool Execute()
+        {
+            EnsureGit();
+
+            Parallel.ForEach(ReadSpecFile(SpecFile), async d =>
+            {
+                var cloneSuccess = await CloneRepo(d.Url, Path.Combine(SourceRoot, d.Name));
+                if (!cloneSuccess)
+                {
+                    Log.LogError($"Cloning {d.Url} into {Path.Combine(SourceRoot, d.Name)} failed.");
+                }
+                var checkoutSuccess = await CheckoutRepo(Path.Combine(SourceRoot, d.Name), d.Hash);
+                if (!checkoutSuccess)
+                {
+                    Log.LogError($"Checking out {d.Hash} in repo {Path.Combine(SourceRoot, d.Name)} failed.");
+                }
+            });
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private async Task<bool> RunExternalGitCommand(string args, string workingDir = null)
+        {
+            workingDir = workingDir ?? Environment.CurrentDirectory;
+            var psi = new ProcessStartInfo
+            {
+                FileName = "git",
+                Arguments = args,
+                CreateNoWindow = true,
+                WorkingDirectory = workingDir,
+                RedirectStandardError = true,
+                RedirectStandardOutput = true,
+            };
+            Log.LogCommandLine($"{psi.FileName} {psi.Arguments}");
+            var p = Process.Start(psi);
+            p.WaitForExit();
```

Comment:
This looks like the classic hang trap to me: https://stackoverflow.com/q/139593. Annoying how easy it is to code it this way.

Maybe reuse MSBuild's Exec task instead of implementing it yourself? I did it that way here: https://github.com/dotnet/buildtools/blob/29c184401b9d65166475e5630ca8459246cfb4d6/src/Microsoft.DotNet.Build.Tasks/ExecWithRetriesForNuGetPush.cs#L100-L112